File: backend/repositories/repositories/submission_repository.py
"""
Submission Repository
Part of the Data Access Layer
Handles database operations for assignment submissions
"""

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, update
from sqlalchemy.orm import selectinload
from repositories.models.submission import Submission
from repositories.models.assignment import Assignment
from repositories.models.user import User
from typing import List, Optional


class SubmissionRepository:
    """Repository for submission-related database operations"""

    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_submission(
        self,
        assignment_id: int,
        student_id: int,
        file_name: str,
        file_path: str,
        file_size: int
    ) -> Submission:
        """
        Create a new submission
        Automatically calculates submission_number and marks it as latest

        Args:
            assignment_id: ID of the assignment
            student_id: ID of the student
            file_name: Name of the uploaded file
            file_path: Path where the file is stored
            file_size: Size of the file in bytes

        Returns:
            Created Submission object
        """
        # Get the next submission number
        existing_submissions = await self.get_submission_history(
            assignment_id, student_id
        )
        submission_number = len(existing_submissions) + 1

        # Mark all previous submissions as not latest
        if existing_submissions:
            await self._update_is_latest_flags(assignment_id, student_id, False)

        # Create new submission
        submission = Submission(
            assignment_id=assignment_id,
            student_id=student_id,
            file_name=file_name,
            file_path=file_path,
            file_size=file_size,
            submission_number=submission_number,
            is_latest=True
        )

        self.db.add(submission)
        await self.db.commit()
        await self.db.refresh(submission)

        return submission

    async def get_submission_by_id(self, submission_id: int) -> Optional[Submission]:
        """
        Get a submission by ID

        Args:
            submission_id: ID of the submission

        Returns:
            Submission object or None if not found
        """
        result = await self.db.execute(
            select(Submission)
            .options(
                selectinload(Submission.assignment),
                selectinload(Submission.student)
            )
            .where(Submission.id == submission_id)
        )
        return result.scalar_one_or_none()

    async def get_submissions_by_assignment(
        self,
        assignment_id: int,
        latest_only: bool = True
    ) -> List[Submission]:
        """
        Get all submissions for an assignment

        Args:
            assignment_id: ID of the assignment
            latest_only: If True, only return the latest submission for each student

        Returns:
            List of Submission objects
        """
        query = (
            select(Submission)
            .options(
                selectinload(Submission.student)
            )
            .where(Submission.assignment_id == assignment_id)
        )

        if latest_only:
            query = query.where(Submission.is_latest == True)

        query = query.order_by(Submission.submitted_at.desc())
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def get_submissions_by_student(
        self,
        student_id: int,
        latest_only: bool = True
    ) -> List[Submission]:
        """
        Get all submissions by a student

        Args:
            student_id: ID of the student
            latest_only: If True, only return the latest submission for each assignment

        Returns:
            List of Submission objects
        """
        query = (
            select(Submission)
            .options(
                selectinload(Submission.assignment)
            )
            .where(Submission.student_id == student_id)
        )

        if latest_only:
            query = query.where(Submission.is_latest == True)

        query = query.order_by(Submission.submitted_at.desc())
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def get_latest_submission(
        self,
        assignment_id: int,
        student_id: int
    ) -> Optional[Submission]:
        """
        Get the latest submission for a specific student-assignment pair

        Args:
            assignment_id: ID of the assignment
            student_id: ID of the student

        Returns:
            Latest Submission object or None if not found
        """
        result = await self.db.execute(
            select(Submission)
            .options(
                selectinload(Submission.assignment),
                selectinload(Submission.student)
            )
            .where(
                and_(
                    Submission.assignment_id == assignment_id,
                    Submission.student_id == student_id,
                    Submission.is_latest == True
                )
            )
        )
        return result.scalar_one_or_none()

    async def get_submission_history(
        self,
        assignment_id: int,
        student_id: int
    ) -> List[Submission]:
        """
        Get all submissions for a specific student-assignment pair (submission history)

        Args:
            assignment_id: ID of the assignment
            student_id: ID of the student

        Returns:
            List of Submission objects ordered by submission number
        """
        result = await self.db.execute(
            select(Submission)
            .where(
                and_(
                    Submission.assignment_id == assignment_id,
                    Submission.student_id == student_id
                )
            )
            .order_by(Submission.submission_number.asc())
        )
        return list(result.scalars().all())

    async def check_submission_exists(
        self,
        assignment_id: int,
        student_id: int
    ) -> bool:
        """
        Check if a student has submitted for an assignment

        Args:
            assignment_id: ID of the assignment
            student_id: ID of the student

        Returns:
            True if submission exists, False otherwise
        """
        result = await self.db.execute(
            select(Submission).where(
                and_(
                    Submission.assignment_id == assignment_id,
                    Submission.student_id == student_id
                )
            )
        )
        return result.scalar_one_or_none() is not None

    async def _update_is_latest_flags(
        self,
        assignment_id: int,
        student_id: int,
        is_latest: bool
    ) -> None:
        """
        Update is_latest flag for all submissions of a student-assignment pair

        Args:
            assignment_id: ID of the assignment
            student_id: ID of the student
            is_latest: New value for is_latest flag
        """
        await self.db.execute(
            update(Submission)
            .where(
                and_(
                    Submission.assignment_id == assignment_id,
                    Submission.student_id == student_id
                )
            )
            .values(is_latest=is_latest)
        )
        await self.db.commit()

File: backend-ts/src/repositories/submission.repository.ts
import { eq, and, desc, sql, inArray } from "drizzle-orm"
import {
  submissions,
  users,
  type Submission,
  type NewSubmission,
} from "@/models/index.js"
import { BaseRepository } from "@/repositories/base.repository.js"
import { injectable } from "tsyringe"

/** Submission with student information */
export interface SubmissionWithStudent {
  submission: Submission
  studentName: string
}

/**
 * Repository for submission-related database operations.
 */
@injectable()
export class SubmissionRepository extends BaseRepository<
  typeof submissions,
  Submission,
  NewSubmission
> {
  constructor() {
    super(submissions)
  }

  /** Get a submission by ID */
  async getSubmissionById(
    submissionId: number,
  ): Promise<Submission | undefined> {
    return await this.findById(submissionId)
  }

  /** Get all submissions for an assignment */
  async getSubmissionsByAssignment(
    assignmentId: number,
    latestOnly: boolean = true,
  ): Promise<Submission[]> {
    if (latestOnly) {
      return await this.db
        .select()
        .from(submissions)
        .where(
          and(
            eq(submissions.assignmentId, assignmentId),
            eq(submissions.isLatest, true),
          ),
        )
        .orderBy(desc(submissions.submittedAt))
    }

    return await this.db
      .select()
      .from(submissions)
      .where(eq(submissions.assignmentId, assignmentId))
      .orderBy(desc(submissions.submittedAt))
  }

  /** Get all submissions by a student */
  async getSubmissionsByStudent(
    studentId: number,
    latestOnly: boolean = true,
  ): Promise<Submission[]> {
    if (latestOnly) {
      return await this.db
        .select()
        .from(submissions)
        .where(
          and(
            eq(submissions.studentId, studentId),
            eq(submissions.isLatest, true),
          ),
        )
        .orderBy(desc(submissions.submittedAt))
    }

    return await this.db
      .select()
      .from(submissions)
      .where(eq(submissions.studentId, studentId))
      .orderBy(desc(submissions.submittedAt))
  }

  /**
   * Get all submissions for a class (via assignments).
   * Used for cleanup during class deletion.
   */
  async getSubmissionsByClass(classId: number): Promise<Submission[]> {
    // We need to import assignments model to join
    const { assignments } = await import("@/models/index.js")

    return await this.db
      .select({
        id: submissions.id,
        assignmentId: submissions.assignmentId,
        studentId: submissions.studentId,
        fileName: submissions.fileName,
        filePath: submissions.filePath,
        fileSize: submissions.fileSize,
        submissionNumber: submissions.submissionNumber,
        submittedAt: submissions.submittedAt,
        isLatest: submissions.isLatest,
        grade: submissions.grade,
        isLate: submissions.isLate,
        penaltyApplied: submissions.penaltyApplied,
        isGradeOverridden: submissions.isGradeOverridden,
        overrideFeedback: submissions.overrideFeedback,
        overriddenAt: submissions.overriddenAt,
      })
      .from(submissions)
      .innerJoin(assignments, eq(submissions.assignmentId, assignments.id))
      .where(eq(assignments.classId, classId))
  }

  /** Get submission history for a student-assignment pair */
  async getSubmissionHistory(
    assignmentId: number,
    studentId: number,
  ): Promise<Submission[]> {
    return await this.db
      .select()
      .from(submissions)
      .where(
        and(
          eq(submissions.assignmentId, assignmentId),
          eq(submissions.studentId, studentId),
        ),
      )
      .orderBy(submissions.submissionNumber)
  }

  /** Get the latest submission for a student-assignment pair */
  async getLatestSubmission(
    assignmentId: number,
    studentId: number,
  ): Promise<Submission | undefined> {
    const results = await this.db
      .select()
      .from(submissions)
      .where(
        and(
          eq(submissions.assignmentId, assignmentId),
          eq(submissions.studentId, studentId),
          eq(submissions.isLatest, true),
        ),
      )
      .limit(1)

    return results[0]
  }

  /** Get submission count for a student-assignment pair */
  async getSubmissionCount(
    assignmentId: number,
    studentId: number,
  ): Promise<number> {
    const result = await this.db
      .select({ count: sql<number>`count(*)` })
      .from(submissions)
      .where(
        and(
          eq(submissions.assignmentId, assignmentId),
          eq(submissions.studentId, studentId),
        ),
      )

    return Number(result[0]?.count ?? 0)
  }

  /** Create a new submission */
  async createSubmission(data: {
    assignmentId: number
    studentId: number
    fileName: string
    filePath: string
    fileSize: number
    submissionNumber: number
    isLate: boolean
    penaltyApplied: number
  }): Promise<Submission> {
    return await this.db.transaction(async (tx) => {
      // Lock relevant rows to serialize concurrent submissions
      await tx
        .select()
        .from(submissions)
        .where(
          and(
            eq(submissions.assignmentId, data.assignmentId),
            eq(submissions.studentId, data.studentId),
          ),
        )
        .for("update")

      // Mark previous submission as not latest
      await tx
        .update(submissions)
        .set({ isLatest: false })
        .where(
          and(
            eq(submissions.assignmentId, data.assignmentId),
            eq(submissions.studentId, data.studentId),
          ),
        )

      // Create new submission
      const results = await tx
        .insert(submissions)
        .values({
          assignmentId: data.assignmentId,
          studentId: data.studentId,
          fileName: data.fileName,
          filePath: data.filePath,
          fileSize: data.fileSize,
          submissionNumber: data.submissionNumber,
          isLatest: true,
          isLate: data.isLate,
          penaltyApplied: data.penaltyApplied,
        })
        .returning()

      // Defensive check to ensure insert succeeded
      if (!results || results.length === 0) {
        throw new Error(
          "Failed to create submission: insert returned no results",
        )
      }

      return results[0]
    })
  }

  /** Get submissions with student info for an assignment */
  async getSubmissionsWithStudentInfo(
    assignmentId: number,
    latestOnly: boolean = true,
  ): Promise<SubmissionWithStudent[]> {
    const query = this.db
      .select({
        submission: submissions,
        studentName: sql<string>`concat(${users.firstName}, ' ', ${users.lastName})`,
      })
      .from(submissions)
      .innerJoin(users, eq(submissions.studentId, users.id))
      .where(
        latestOnly
          ? and(
            eq(submissions.assignmentId, assignmentId),
            eq(submissions.isLatest, true),
          )
          : eq(submissions.assignmentId, assignmentId),
      )
      .orderBy(desc(submissions.submittedAt))

    return await query
  }

  /** Get a single submission with student name */
  async getSubmissionWithStudent(
    submissionId: number,
  ): Promise<SubmissionWithStudent | null> {
    const results = await this.db
      .select({
        submission: submissions,
        studentName: sql<string>`concat(${users.firstName}, ' ', ${users.lastName})`,
      })
      .from(submissions)
      .innerJoin(users, eq(submissions.studentId, users.id))
      .where(eq(submissions.id, submissionId))
      .limit(1)

    return results[0] ?? null
  }

  /** Get multiple submissions with student names by IDs */
  async getBatchSubmissionsWithStudents(
    submissionIds: number[],
  ): Promise<SubmissionWithStudent[]> {
    if (submissionIds.length === 0) return []

    return await this.db
      .select({
        submission: submissions,
        studentName: sql<string>`concat(${users.firstName}, ' ', ${users.lastName})`,
      })
      .from(submissions)
      .innerJoin(users, eq(submissions.studentId, users.id))
      .where(inArray(submissions.id, submissionIds))
  }

  /**
   * Get total submission count.
   * Used for admin analytics dashboard.
   */
  async getTotalCount(): Promise<number> {
    const result = await this.db
      .select({ count: sql<number>`count(*)` })
      .from(submissions)
    return Number(result[0]?.count ?? 0)
  }

  /**
   * Update the grade for a submission.
   * Called after test execution to store the calculated grade.
   */
  async updateGrade(submissionId: number, grade: number): Promise<void> {
    await this.db
      .update(submissions)
      .set({ grade })
      .where(eq(submissions.id, submissionId))
  }

  /**
   * Set a grade override for a submission.
   * Updates the grade, marks it as overridden, and optionally adds feedback.
   */
  async setGradeOverride(
    submissionId: number,
    grade: number,
    feedback: string | null,
  ): Promise<void> {
    await this.db
      .update(submissions)
      .set({
        grade,
        isGradeOverridden: true,
        overrideFeedback: feedback,
        overriddenAt: new Date(),
      })
      .where(eq(submissions.id, submissionId))
  }

  /**
   * Remove a grade override from a submission.
   * Clears the override flag and feedback. The grade should be recalculated separately.
   */
  async removeGradeOverride(submissionId: number): Promise<void> {
    await this.db
      .update(submissions)
      .set({
        isGradeOverridden: false,
        overrideFeedback: null,
        overriddenAt: null,
      })
      .where(eq(submissions.id, submissionId))
  }
}
